app/models/miq_expression.rb

```diff
@@ -566,6 +566,10 @@ def self._to_ruby(exp, context_type, tz)
     clause
   end
 
+  def normalize_date_time(rel_time, tz, mode = "beginning")
```

Comment:
Perhaps some (if not all) of these new methods could be made private? I think it would be a good idea to reduce the surface area of this large class where possible, and make a stronger distinction over what its public interface is.